// Dune/View/Caption.js
module.exports = new Caption();

var canvasContainer = require("Dune/CanvasContainer");

function Caption() 
{

  var canvas, context;

  this.draw = function(captionText) 
  {
    if (! captionText || ! captionText.length)
      throw new Error("No text argument passed");

    canvas = canvasContainer.layer('notification');
    context = canvas.getContext("2d");

    var captionTextAttr = getCaptionTextAttr(captionText);
    var captionBorderDimensions = getCaptionBorderDimensions(captionTextAttr);

    drawCaptionBorder(captionBorderDimensions);
    drawCaptionText(captionTextAttr, captionBorderDimensions);
  };

  function getCaptionTextAttr(text) 
  {
    var fontSize = 30;
    context.font = fontSize+"pt Arial";

    return {
      "text": text,
      "padding": 20,
      "fontSize": fontSize,
      "font": fontSize+"pt Arial",
      "width": context.measureText(text).width,
      "height": fontSize
    };

  }

  function getCaptionBorderDimensions(captionTextAttr) {
    return { 
      "buffer1": 8,
      "buffer2": 3.6,
      "width": captionTextAttr.width + captionTextAttr.padding,
      "height": captionTextAttr.fontSize * 2,
      "x": canvas.width/2 - captionTextAttr.width/2 - captionTextAttr.padding/2,
      "y": 20
    };
  }

  function drawCaptionBorder(border) {
    context.beginPath();

    drawCaptionBorderBottomRight(border);
    drawCaptionBorderBottomLeft(border);
    drawCaptionBorderTopLeft(border);
    drawCaptionBorderTopRight(border);

    context.closePath();

    var transparentBlackStyle = "rgba(20, 20, 20, 0.6)";
    context.fillStyle = transparentBlackStyle;
    context.fill();
    context.lineWidth = 2.0;
    context.strokeStyle = "rgb(255, 255, 255)";
    context.stroke();

  }

  function drawCaptionBorderBottomRight(border) 
  {
    var borderWidth = border.x + border.width;
    var borderHeight = border.y + border.height;

    var startX = borderWidth;
    var startY = borderHeight - border.buffer1;

    var xMidpoint1 = startX;
    var yMidpoint1 = borderHeight - border.buffer2;

    var xMidpoint2 = borderWidth - border.buffer2; 
    var yMidpoint2 = borderHeight;

    var endX = borderWidth - border.buffer1;
    var endY = yMidpoint2;

    context.moveTo(startX, startY);
    context.lineTo(startX, startY);
    context.bezierCurveTo(
	xMidpoint1, yMidpoint1, 
	xMidpoint2, yMidpoint2, 
	endX, endY);
  }

  function drawCaptionBorderBottomLeft(border)
  {
    var startX = border.x + border.buffer1;
    var startY = border.y + border.height;

    var xMidpoint1 = border.x + border.buffer2;
    var yMidpoint1 = startY;

    var xMidpoint2 = border.x;
    var yMidpoint2 = yMidpoint1;

    var endX = border.x;
    var endY = yMidpoint2 - border.buffer1;

    context.lineTo(startX, startY);
    context.bezierCurveTo(
      xMidpoint1, yMidpoint1,
      xMidpoint2, yMidpoint2,
      endX, endY);
  }

  function drawCaptionBorderTopLeft(border) 
  {
    var startX = border.x;
    var startY = border.y + border.buffer1;

    var xMidpoint1 = startX;
    var yMidpoint1 = border.y + border.buffer2;

    var xMidpoint2 = border.x + border.buffer2;
    var yMidpoint2 = border.y;

    var endX = border.x + border.buffer1;
    var endY = border.y;

    context.lineTo(startX, startY);
    context.bezierCurveTo(
      xMidpoint1, yMidpoint1,
      xMidpoint2, yMidpoint2,
      endX, endY);
  }

  function drawCaptionBorderTopRight(border) 
  {
    var startX = border.x + border.width - border.buffer1;
    var startY = border.y;

    var xMidpoint1 = border.x + border.width - border.buffer2;
    var yMidpoint1 = border.y;

    var xMidpoint2 = border.x + border.width;
    var yMidpoint2 = border.y + border.buffer2;

    var endX = xMidpoint2;
    var endY = border.y + border.buffer1;

    context.lineTo(startX, startY);
    context.bezierCurveTo(
      xMidpoint1, yMidpoint1,
      xMidpoint2, yMidpoint2,
      endX, endY);
  }

  function drawCaptionText(captionTextAttr, captionBorderDimensions) {

    var horizontalCenter = canvas.width/2;

    var xPos = horizontalCenter - captionTextAttr.width/2;
    var yPos = captionBorderDimensions.height + captionBorderDimensions.y/4;

    context.fillStyle = "white";
    context.fillText(captionTextAttr.text, xPos, yPos);
  }

}
