// Dune/View/TraitorSelect.js
//module.exports = promptUserSelectTraitor;
module.exports = TraitorSelect;

var canvasContainer = require("Dune/CanvasContainer");
var Loader = require("Dune/Loader");
var eventChain = require("Dune/EventChain");
var caption = require("Dune/View/Caption");

var canvas, context, canvasContainer;

var traitorScaleWidth = 275,
    traitorScaleHeight = 400;


var factionView;

function TraitorSelect() {
  
  this.promptUser = function(view) {

    factionView = view;

    canvas = canvasContainer.layer('notification');

    context = canvas.getContext("2d");

    var loader = new Loader();

    faction = factionView.faction;

    canvas.elements = new Array();

    var that = this;

    loadTraitorHandImages();
    addOnClickEvent();

    function loadTraitorHandImages() {
      var traitorHand = faction.drawTraitorHand();

      for (var i = 0; i < traitorHand.length; i++) {
	var traitor = traitorHand[i];
	var factionName = traitor.faction.toLowerCase().split(' ').join('');

	var imageUrl = "/img/traitors/" + factionName + "/" 
	  + traitor.name.toLowerCase().split(' ').join('_') + ".png";

	var image = loader.loadImage(imageUrl);
	image.traitor = traitor;
	canvas.elements[i] = image;
      }
    }

    function addOnClickEvent() {
      canvas.addEventListener('click', function(element) {
      	that.selectTraitor(this, element);

      });
    }


    loader.onload = drawTraitorSelectScreen;
  }

  this.selectTraitorImage = function(image)
  {
    factionView.takeCard(image);

    var traitorObj = image.traitor;
    var faction = factionView.faction;
    faction.pickTraitor(traitorObj);

    canvasContainer.deleteLayer(canvas);

    eventChain.next();
  }

}

TraitorSelect.prototype.selectTraitor = function(canvas, element) {
  var coord = getMousePosition(canvas,element);

  for (var i = 0; i < canvas.elements.length; i++) {
    var image = canvas.elements[i];

    if (coord.x >= image.xPos && coord.x <= image.xPos + traitorScaleWidth
	&& coord.y >= image.yPos && coord.y <= image.yPos + traitorScaleHeight
    ) {
      this.selectTraitorImage(image);
    }
  }
}

function getMousePosition(canvasElement,e)
{
  var rect = canvasElement.getBoundingClientRect();
  var mousex = e.clientX - rect.left; 
  var mousey = e.clientY - rect.top;

  return {x: mousex, y: mousey};
}



function drawTraitorSelectScreen() 
{
  var captionText = "Choose your traitor";
  caption.draw(captionText);
  drawTraitors();
}

function drawTraitors()
{

  var image1 = canvas.elements[0];

  image1.xPos = 75;
  image1.yPos = 100;

  var image2 = canvas.elements[1];

  image2.xPos = canvas.width - traitorScaleWidth - image1.xPos;
  image2.yPos = image1.yPos;

  var image3 = canvas.elements[2];

  image3.xPos = image1.xPos;
  image3.yPos = image2.yPos + image2.xPos - traitorScaleWidth - image1.xPos + traitorScaleHeight ;

  var image4 = canvas.elements[3];

  image4.xPos = image2.xPos;
  image4.yPos = image3.yPos;

  context.drawImage(
      image1, image1.xPos, image1.yPos, traitorScaleWidth, traitorScaleHeight);

  context.drawImage(
      image2, image2.xPos, image2.yPos, traitorScaleWidth, traitorScaleHeight);

  context.drawImage(
      image3, image3.xPos, image3.yPos, traitorScaleWidth, traitorScaleHeight);

  context.drawImage(
      image4, image4.xPos, image4.yPos, traitorScaleWidth, traitorScaleHeight);
}
