// Dune/CanvasContainer.js
module.exports = new CanvasContainer;

function CanvasContainer() 
{

  var container = document.getElementById("gamecontainer");

  var layerMap = { };
  
  this.layer = function(layerName) {
    if (! layerMap[layerName]) 
      newLayer(layerName);

    return layerMap[layerName]
  }

  function newLayer(layerName) {
    var canvas = document.createElement("canvas");

    layerMap[layerName] = canvas;
    setCanvasAttributes(canvas, layerName);

    container.appendChild(canvas);
  }

  function setCanvasAttributes(canvas, layerName) {
    var zIndex = zIndexLookup(layerName);

    canvas.id = layerName;
    canvas.className = "gamelayer";
    canvas.style.display = "block";
    canvas.style.zIndex = zIndex;
    canvas.width = 768;
    canvas.height = 1024;

    if (layerName == "test")
      canvas.style.display = "none";

  }

  function zIndexLookup(layerName) 
  {
    switch (layerName) {
      case "notification":
      	return 200;
      case "playerscreen":
      	return 99;
      case "troopscreen":
      	return 98;
      case "playerseat":
      	return 97;
      case "storm":
      	return 96;
      case "test":
      	return 95;
      default:
      	return 99;
              //throw new Error("No zIndex case defined for " + layerName);
    }
  }

  this.deleteLayer = function(canvas) 
  {
    delete layerMap[canvas.id];
    if (canvas.parentNode === container)
      container.removeChild(canvas);
  }
}
