// Dune/Controller.js
module.exports = GameController;

var DuneGame = require("Dune/Game");
var MapView = require("./View/Map");
var StartMenuView = require("./View/StartMenu");
var FactionSelectView = require("./View/FactionSelect");

function GameController() {

  var factions = [];
  var views = {
    "map": new MapView(this),
    "start": new StartMenuView(this),
    "select":new FactionSelectView(this),
    "factions": []
  };
  var turn = 0;
  var traitorPool = new Array();
  var that = this;

  this.game = new DuneGame();

  //getStartMenuView();

/*  function getStartMenuView() {*/
    //if (! views.startMenu) {
    //}

    //return views.startMenu

  /*}*/

  this.startGame = function() {

    views.select.hide();
    //hideFactionSelectView();
    //initNewGame();
    initFactionViews();
    this.game.start();
    initMapView();
    //shuffleTurnOrder();
    //makeTraitorPool();
    startInitialTurn();
  }

/*  this.hideStartMenuView = function() {*/
    //var startMenuView = getStartMenuView();
    //startMenuView.hide();
  //}

  //this.showFactionSelectView = function() {
    //var factionSelectView = this.getFactionSelectView();
    //factionSelectView.show();
  /*}*/

  function initFactionViews() {
    var factionViews = new Array();

    for (var i = 0; i < factions.length; i++) {
      var FactionView = getFactionViewConstructor(factions[i]);
      views.factions.push(new FactionView(that));
    }

    setFactionViews(factionViews);
  }

  function getFactionViewConstructor(factionName) {
    switch(factionName) {
      case "Atreides":
      	return require("./View/Faction/Atreides");
      case "Harkonnen":
      	return require("./View/Faction/Harkonnen");
      case "BeneGesserit":
      	return require("./View/Faction/BeneGesserit");
      case "Fremen":
      	return require("./View/Faction/Fremen");
      case "Guild":
      	return require("./View/Faction/Guild");
      case "Emperor":
      	return require("./View/Faction/Emperor");
      default:
      	throw new Error("Invalid faction view: "+factionName);
    }
  }

  function setFactionViews(array) {
    views.factions = array;
  }

  function initMapView() {
    views.map.show();
    views.map.drawStorm();
  }

  this.debugSetFactions = function (factionsArray) {
  /* This is a debug function and should go away once gameplay finalized */
    factions = factionsArray;
  }

  function startInitialTurn() {
    //console.log('starting initial turn');
    //makeTraitorPool();
    //var factionView = getNextTurnOrder();
    //factionView.startInitialTurn();
  }

}
