// Dune/Map.js
module.exports = ArrakisMap;

function ArrakisMap() {
  var my = this;
  var locations = {};

  my.getLocation = function(locationName) {

    if (! locations[locationName]) 
      locations[locationName] = getArrakisMapLocation(locationName);

    return locations[locationName];
  }


  return my;
}

function getArrakisMapLocation(locationName) {

  switch(locationName) {
    case "PolarSink":
      return new PolarSink();
  }
}

function PolarSink() {
  var my = this;
  var factions = [];

  var neighbors = [
    WindPass, WindPassNorth, CielagoNorth, HargPass, FalseWallEast, 
    ImperialBasin, Arsunt, HaggaBasin
  ];

  my.getNeighbors = function() {
    return neighbors.length;
  };

  my.getFactionCount = function() {
    return factions.length;
  }

  my.getFactions = function() {
    return factions;
  };

  my.occupy = function(troops) {
    factions.push(troops);
  }

  return my;
}

function WindPass() { }
function WindPassNorth() { }
function CielagoNorth() { }
function HargPass() { }
function FalseWallEast() { }
function ImperialBasin() { }
function Arsunt() { }
function HaggaBasin() { }
