// Dune/Map/Base.js
module.exports = BaseTerritory;

function BaseTerritory() {

  var neighbors = [];

  var factions = {};

  this.getNeighbors = function() {
    _initNeighbors(this);
    return neighbors;
  }

  this.getNeighbor = function(neighborName) {
    _initNeighbors(this);

    for (var i = 0; i < neighbors.length; i++) {
      var neighbor = neighbors[i];
      if (neighbor.constructor.name === neighborName) 
      return neighbor;

    }

    throw new Error(
      "Invalid neighbor of " + this.constructor.name + ": " + neighborName 
    );

  };

  function _initNeighbors(that) {
  /* Lazy load function. Set neighbors only when we need them */
    if (! neighbors.length) that.setNeighbors();
  }

  this.setNeighbors = _trapNeighborListInTerritoryFunctionScope;

  function _trapNeighborListInTerritoryFunctionScope(neighborList) {
    /* Give territory neighbor list access without public exposing it */
    var that = this;
    this.setNeighbors = function() {
      neighbors = _convertNeighborListToTerritories(that, neighborList);
    }

  };

  function _convertNeighborListToTerritories(that, neighborList) {
    var newNeighbors = [];
    var map = that.getMap();

    for (var i = 0; i < neighborList.length; i++) {
      var neighbor = neighborList[i];
      var territory = map.getTerritory(neighbor);

      newNeighbors.push(territory);
    }

    return newNeighbors;
  }

  this.getFactions = function() {
    return factions;
  };

  this.getFactionSize = function() {
    return Object.keys(factions).length 
  }

  this.clearFactions = function() {
    factions = {};
  }

  this.addTroops = function(troops) {
    var faction = troops.getFaction();

    if (factions[faction]) {
      factions[faction].push(troops);
    } else {
      factions[faction] = new Array(troops);
    }

    if (this.getFactionSize() > 1) {
      var map = this.getMap();
      map.addConflict(this);
    }

  }

  return this;
}
