// Dune/View/Deck/Bonus.js
var Loader = require("Dune/Loader");
var canvasContainer = require("Dune/CanvasContainer");
var mapView = require("Dune/View/Map");
var eventChain = require("Dune/EventChain");

module.exports = new BonusDeckView();

function BonusDeckView() 
{
  var canvas;

  var bonusDeckImages = new Array(
    "carryalls", "harvesters", "ornithopters", "smugglers"
  );

  var loader = new Loader();

  var cardScaleWidth = 333,
      cardScaleHeight = 483;

  var mapDimensions = mapView.circle;

  var xCenterScreen = mapDimensions.centerX - cardScaleWidth/2;
  var yCenterScreen = mapDimensions.centerY - cardScaleHeight/2;

  for (var i = 0; i < bonusDeckImages.length; i++) {
    var bonusDeckImage = bonusDeckImages[i];
    var bonusImgUrl = "/img/bonus/" + bonusDeckImage + ".png";
    bonusDeckImages[bonusDeckImage] = loader.loadImage(bonusImgUrl);
  }

  this.dealCard = function(cardName) 
  {
    canvas = canvasContainer.layer("notification");
    var bonusCard = getBonusCardImage(cardName);

    moveDealtCardCardFromOffscreenRightToCenter(bonusCard);
    pauseBrieflyThenMoveDealtCardLeftUntilOffscreen(bonusCard);

    return bonusCard;

  }


  function moveDealtCardCardFromOffscreenRightToCenter(bonusCardImg) {
    context = canvas.getContext("2d");

    canvas.redraw = function() { 
      context.clearRect(0, 0, canvas.width, canvas.height)
    }
    bonusCardImg.moveToCoord([xCenterScreen, yCenterScreen]);
  }

  function pauseBrieflyThenMoveDealtCardLeftUntilOffscreen(bonusCardImg) {
    bonusCardImg.onHalt = function() {
      var timeout = 1000;
      //DEBUG
      timeout = 0;

      setTimeout(function() { 

	var xPos = 0 - cardScaleWidth;
	var yPos = yCenterScreen;

	bonusCardImg.moveToCoord([xPos, yPos]);

	bonusCardImg.onHalt = function() {
	  canvasContainer.deleteLayer(canvas);
	  eventChain.next();
	}
      }, timeout);
    }
  }

  function getBonusCardImage(cardName) {

    if (!bonusDeckImages[cardName]) 
      throw new Error("No bonus card named " + cardName);

    var bonusCardImg = bonusDeckImages[cardName];

    var xPos = mapDimensions.centerX - cardScaleWidth/2;
    var yPos = mapDimensions.centerY - cardScaleHeight/2;

    bonusCardImg.canvas = canvas;
    bonusCardImg.xPos = canvas.width;
    bonusCardImg.yPos = yPos;
    bonusCardImg.speed = 0.02;
    bonusCardImg.width = cardScaleWidth;
    bonusCardImg.height = cardScaleHeight;

    return bonusCardImg
  }

}
