// Dune/View/Deck/Treachery.js
var mapView = require("Dune/View/Map");
var Loader = require("Dune/Loader");
var canvasContainer = require("Dune/CanvasContainer");
var shuffleArray = require("Dune/shuffle");
var eventChain = require("Dune/EventChain");

module.exports = new TreacheryDeckView();

function TreacheryDeckView() 
{
  var canvas,context;

  var loader = new Loader();

  var cardScaleWidth = 333,
      cardScaleHeight = 483;

  var mapDimensions = mapView.circle;

  var xCenterScreen = mapDimensions.centerX - cardScaleWidth/2;
  var yCenterScreen = mapDimensions.centerY - cardScaleHeight/2;

  var treacheryDeckImages = new Array(
    "baliset", "chaumas", "chaumurky", "chrysknife", "cheapheroine", 
    "ellacadrug", "familyatomics", "gomjabbar", "hajr", 
    "harvester", "jubbacloak", "kulon", "lalala", "lasgun", 
    "maulapistol", "sliptip", "stunner", "thumper", "tleilaxughola", 
    "triptogamont", "weathercontrol"
  );

  var duplicateCards = new Array(
    Array.apply(null, new Array(4)).map(function() { return "shield" }),
    Array.apply(null, new Array(4)).map(function() { return "snooper" }),
    Array.apply(null, new Array(2)).map(function() { return "cheaphero" }),
    Array.apply(null, new Array(2)).map(function() { return "karama" })
  );

  for (var i = 0; i < duplicateCards.length; i++) {
    var duplicateCardArray = duplicateCards[i];
    treacheryDeckImages = treacheryDeckImages.concat(duplicateCardArray);
  }

  for (var i = 0; i < treacheryDeckImages.length; i++) {
    var treacheryDeckImage = treacheryDeckImages[i];
    var treacheryImgUrl = "/img/treachery/" + treacheryDeckImage + ".png";
    treacheryDeckImages[i] = loader.loadImage(treacheryImgUrl);
  }

  loader.onload = function() {
    shuffleArray(treacheryDeckImages);
  }


  var that = this;

  this.dealCard = function() {

    canvas = canvasContainer.layer("notification");
    var treacheryImg = getNewTreacheryImage();

    moveDealtCardCardFromOffscreenRightToCenter(treacheryImg);
    pauseBrieflyThenMoveDealtCardLeftUntilOffscreen(treacheryImg);

    return treacheryImg;
  }

  function moveDealtCardCardFromOffscreenRightToCenter(treacheryImg) {
    context = canvas.getContext("2d");

    canvas.redraw = function() { 
      context.clearRect(0, 0, canvas.width, canvas.height)
    }
    treacheryImg.moveToCoord([xCenterScreen, yCenterScreen]);
  }

  function pauseBrieflyThenMoveDealtCardLeftUntilOffscreen(treacheryImg) {
    treacheryImg.onHalt = function() {
      var timeout = 1000;
      //DEBUG
      timeout = 0;

      setTimeout(function() { 

	var xPos = 0 - cardScaleWidth;
	var yPos = yCenterScreen;

	treacheryImg.moveToCoord([xPos, yPos]);

	treacheryImg.onHalt = function() {
	  canvasContainer.deleteLayer(canvas);
	  eventChain.next();
	}
      }, timeout);
    }
  }

  function getNewTreacheryImage() {
    var treacheryImg = treacheryDeckImages.shift();


    var xPos = mapDimensions.centerX - cardScaleWidth/2;
    var yPos = mapDimensions.centerY - cardScaleHeight/2;

    treacheryImg.canvas = canvas;
    treacheryImg.xPos = canvas.width;
    treacheryImg.yPos = yPos;
    treacheryImg.speed = 0.02;
    treacheryImg.width = cardScaleWidth;
    treacheryImg.height = cardScaleHeight;

    return treacheryImg
  }
}
