// Dune/Factions/Base.js
module.exports = BaseFaction;

function BaseFaction() {
  this.name = "Base";

  var spice = 0;
  var troops = [];

  assignTroops(troops);

  this.getTroopSize = getTroopSize;
  this.getTroops = getTroops;
  this.getSpice = getSpice;
  this.setSpice = setSpice;

  
  function getTroopSize() {
    return troops.length;
  }

  function getTroops(count) {
    if (count > troops.length) 
      throw Error("Not enough troops");

    var container = new BaseTroopContainer(this, troops.slice(0,count));
    return container;
  };

  function getSpice() {
    return spice;
  };

  function setSpice(value) {
    spice = value;
  }

  return this;
}

function assignTroops(troopArray) {
  for (var i = 0; i < 20; i++) {
    troopArray.push(new BaseFactionTroop());
  }
}

function BaseTroopContainer(faction, troops) {

  var faction = faction.constructor.name;
  var currentLocation = undefined;

  this.getFaction = function() {
    return faction;
  }

  this.getSize = function() {
    return troops.length
  }

  this.getContents = function() {
    return troops;
  }

  this.occupy = function(territory) {
    if (currentLocation !== territory) {
      currentLocation = territory;
      territory.addTroops(this);
    }
  }

  return this;
}

function BaseFactionTroop() {
  var strength = 0;

  this.getStrength = function() {
    return strength;
  }

  return this;
}
