// Dune/Factions/Base.js
module.exports = BaseFaction;

function BaseFaction() {
  var my = this;
  my.name = "Base";

  var spice = 0;
  var troops = [];
  var currentLocation = undefined;

  assignTroops(troops);
  
  my.getTroopSize = function() {
    return troops.length;
  }

  my.getTroops = function(count) {
    var container = new makeTroopContainer(this, troops.slice(0,count));
    return container;
  };

 function makeTroopContainer(faction, troops) {
    var faction = faction.constructor.name;

    this.getFaction = function() {
      return faction;
    }

    this.getSize = function() {
      return troops.length
    }

    this.mergeWith = function(newTroops) {
      troops = troops.concat(newTroops.removeContents());
    }

    this.getContents = function() {
      return troops;
    }

    this.clearContents = function() {
      troops = [];
    }

    this.removeContents = function() {
      var contents = this.getContents();
      this.clearContents();
      return contents;
    }

    return this;
  }
 
  my.getSpice = function() {
    return spice;
  };

  my.setSpice = function(value) {
    spice = value;
  }

  my.occupy = function(territory) {
    if (currentLocation !== territory) {

      currentLocation = territory;
    }
  }

  return my;
}

function assignTroops(troopArray) {
  for (var i = 0; i < 20; i++) {
    troopArray.push(new BaseFactionTroop());
  }
}



function BaseFactionTroop() {
  var strength = 0;

  this.getStrength = function() {
    return strength;
  }

  return this;
}
