// Dune/Faction/Base.js
module.exports = BaseFaction;

function BaseFaction(obj, props) {

  var name = props.name;
  var spice = props.spice; 
  var leaders = props.leaders; 
  var game = props.game; 
  var troops = props.troops || assignTroops();
  var traitor = undefined;
  var traitorHand = undefined;
  var handLimit = props.handLimit || 4;
  var treacheryHand = [];

  setEachLeaderFaction();

  function setEachLeaderFaction() 
  {
    for (var i = 0; i < leaders.length; i++) {
      var leader = leaders[i];
      leader.faction = obj.constructor.name;
    }
  }

  function getTroopSize() {
    return troops.length;
  }

  obj.getTroops = function (count) {
    if (count > troops.length) 
      throw Error("Not enough troops");

    var container = new BaseTroopContainer(this, troops.slice(0,count));
    return container;
  };

  obj.leaders = function() {
    return leaders;
  }

  obj.spice = function() {
    return spice;
  }

  obj.drawTraitorHand = function () {
    return traitorHand = game.dealTraitorHand();
  }

  obj.pickTraitor = function(newTraitor) {
    return traitor = newTraitor
  }

  obj.drawTreacheryCard = function() {
    if (obj.isAtHandLimit()) 
      throw new Error("Hand limit reached!");

    var treacheryCard = game.dealTreacheryCard();
    treacheryHand.push(treacheryCard);
  }

  obj.takeTreacheryCard = function(treacheryCard)
  {
    treacheryHand.push(treacheryCard);
  }

  obj.treacheryCardCount = function() { return treacheryHand.length }

  obj.isAtHandLimit = function() {
    return treacheryHand.limit == handLimit
  }

  obj.placeBid = function(spiceBid) 
  {
    if (spiceBid > spice) throw new Error("Invalid bid");
    var bidRound = this.bidRound;
    bidRound.increaseMinimumBid(spiceBid);
  }

  obj.passBid = function() 
  {
    var bidRound = this.bidRound;
    bidRound.passCurrentBidder();
  }

}

function assignTroops(troopArray) {
  var array = new Array();
  for (var i = 0; i < 20; i++) {
    array.push(new BaseFactionTroop());
  }
  return array;
}

function assignLeaders(leaderArray) {
  for (var i = 0; i < 20; i++) {
    leaderArray.push(new BaseFactionTroop());
  }
}

function BaseTroopContainer(faction, troops) {

  var faction = faction.constructor.name;
  var currentLocation = undefined;

  this.getFaction = function() {
    return faction;
  }

  this.getSize = function() {
    return troops.length
  }

  this.getContents = function() {
    return troops;
  }

  this.occupy = function(territory) {
    if (currentLocation !== territory) {
      currentLocation = territory;
      territory.addTroops(this);
    }
  }

  return this;
}

function BaseFactionTroop() {
  var strength = 0.5;

  this.getStrength = function() {
    return strength;
  }

  return this;
}
